cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -592,6 +592,70 @@ struct Utf8ReverseTransform : public StringTransformBase {
 template <typename Type>
 using Utf8Reverse = StringTransformExec<Type, Utf8ReverseTransform>;
 
+struct Utf8NfTransformBase : public StringTransformBase {
+  int64_t MaxCodeunits(int64_t ninputs, int64_t input_ncodeunits) override {
+    // Section 5.18 of the Unicode spec claims that the number of codepoints for case
```

Comment:
Should inherit from [`FunctionalCaseMappingTransform`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_string.cc#L423) which already provides the growth-aware `MaxCodeunits`.

Reply:
Hmm... is the growth factor actually the same for normalization as for case transforms? That seems a bit surprising.

Also, the growth factor could actually depend on the actual normalization: combined forms should arguably be more compact that decomposed form?

Reply:
Answer: there are really pathological cases, so I think it would be better to dynamically resize the output:
```python
>>> s = "\ufdfa"
>>> len(s)
1
>>> len(unicodedata.normalize("NFKD", s))
18
>>> len(unicodedata.normalize("NFD", s))
1
>>> len(unicodedata.normalize("NFKC", s))
18
>>> len(unicodedata.normalize("NFC", s))
1
```